mesonbuild/backend/ninjabackend.py

```diff
@@ -2381,7 +2381,7 @@ def guess_external_link_dependencies(self, linker, target, commands, internal):
 
         build_dir = self.environment.get_build_dir()
         # the following loop sometimes consumes two items from command in one pass
-        it = iter(commands)
+        it = iter(linker.native_link_args_to_unix(commands) if hasattr(linker, 'native_link_args_to_unix') else commands)
```

Comment:
I really dislike adding methods to subclasses and then checking for their existance with `hasattr`, it has a code smell to it. Can we do like we've done for other cases like this and just add this method to the DynamicLinker class that just returns commands unmodified?

Reply:
Sure, although I should mention that despite the name `linker` here is actual a `Compiler` instance. I can still add this method to the `Compiler` or do `linker.linker.native_link_args_to_unix(...)` or bite the bullet and rename `linker` to `compiler'. Let me know which route you prefer.

I went with the attr check mostly because there is another one just a little below (`if hasattr(linker, 'get_library_naming'):`).

Reply:
Add it to `Compiler`. We probably should rename this, but that's another patch.

I've been slowly trying to remove all of those hasattr checks, I just haven't gotten them all yet, obviously :)

Reply:
Ha, I take it back, I guess `linker` can be either a `Compiler` or a `Linker` instance depending on the logic path. Reworked this to match existing `unix_args_to_native`, please take another look.